contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,443 @@
+    * Migration must be modular so that each plugin can have migration turned on
+      and off separately
+
+## Non-Goals
+* Design a mechanism for deploying  CSI drivers on all systems so that users can
+  use the current storage system the same way they do today without having to do
+  extra set up.
+* Implementing CSI Drivers for existing plugins
+* Define set of volume plugins that should be migrated to CSI
+
+## Implementation Schedule
+
+Alpha [1.12]
+* Off by default
+* Proof of concept migration of at least 2 storage plugins [AWS, GCE]
+* Framework for plugin migration built for Dynamic provisioning, pre-provisioned
+  volumes, and in-tree volumes
+
+Beta [Target 1.14]
+* On by default
+* Migrate all of the cloud provider plugins*
+
+GA [TBD]
+* Feature on by default, per-plugin toggle on for relevant cloud provider by
+  default
+* CSI Drivers for migrated plugins available on related cloud provider cluster
+  by default
+
+## Feature Gating
+We will have an alpha feature gate for the whole feature that can turn the CSI
+migration on or off, when off all code paths should revert/stay with the in-tree
+plugins. We will also have individual flags for each driver so that admins can
+toggle them on or off.
+
+The feature gate can exist at the interception points in the OperationGenerator
+for Attach and Mount, as well as in the PV Controller for Provisioning.
+
+We will also have one feature flag for each driver’s migration so that each
+driver migration can be turned on and off individually. 
+
+The new feature gates for alpha are:
+```
+// Enables the in-tree storage to CSI Plugin migration feature.
+CSIMigration utilfeature.Feature = "CSIMigration"
+
+// Enables the GCE PD in-tree driver to GCE CSI Driver migration feature.
+CSIMigrationGCE utilfeature.Feature = "CSIMigrationGCE"
+
+// Enables the AWS in-tree driver to AWS CSI Driver migration feature.
+CSIMigrationAWS utilfeature.Feature = "CSIMigrationAWS"
+```
+
+## Translation Layer
+The main mechanism we will use to migrate plugins is translating In-tree plugin
+sources into CSI Sources and redirecting operation calls to the CSI Driver
+instead of the in-tree driver.
+
+Pros:
+* Keeps old API objects as they are
+* Facilitates gradual roll-over to CSI
+
+Cons:
+* Somewhat complicated and error prone.
+* AttachDetach Controller must have permissions to modify PVs.
+* Bespoke translation logic for each in-tree plugin
+
+### Dynamically Provisioned Volumes
+
+#### Kubernetes Changes
+In-tree provisioner will be changed to have a feature gate that when turned on
+will not register the in-tree volume plugin into PV controller. We will modify
+`provisionClaimOperation` to set the claim provisioner to the migrated external
+provisioner name, so the external provisioner will pick this up and treat it as
+a CSI Volume from then on. The controller will automatically annotate PVCs for
+external provisioning by an provisioner that handles e.g. “google.io/gce-pd”.
```

Comment:
shouldn't provisioning be determined by something on the storage class, not on the PVC? I wouldn't expect any visible change on the PVC